cmd/tealdbg/localLedger.go

```diff
@@ -18,103 +18,103 @@ package main
 
 import (
 	"fmt"
+
+	"github.com/algorand/go-algorand/config"
 	"github.com/algorand/go-algorand/data/basics"
 	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/data/transactions/logic"
+	"github.com/algorand/go-algorand/ledger"
 )
 
-type localLedger struct {
-	round      int
+type balancesAdapter struct {
 	balances   map[basics.Address]basics.AccountData
 	txnGroup   []transactions.SignedTxn
 	groupIndex int
+	proto      config.ConsensusParams
+	round      int
 }
 
-func (l *localLedger) Balance(addr basics.Address) (basics.MicroAlgos, error) {
-	br, ok := l.balances[addr]
-	if !ok {
-		return basics.MicroAlgos{}, fmt.Errorf("no such address %s", addr.String())
+const defaultNewAppIdx = 1380011588
+
+func makeAppLedger(
+	balances map[basics.Address]basics.AccountData, txnGroup []transactions.SignedTxn,
+	groupIndex int, proto config.ConsensusParams, round int,
+) (logic.LedgerForLogic, error) {
+
+	if groupIndex >= len(txnGroup) {
+		return nil, fmt.Errorf("invalid groupIndex %d exceed txn group length %d", groupIndex, len(txnGroup))
 	}
-	return br.MicroAlgos, nil
-}
+	txn := txnGroup[groupIndex]
 
-func (l *localLedger) Round() basics.Round {
-	return basics.Round(l.round)
-}
+	accounts := []basics.Address{txn.Txn.Sender}
+	for _, addr := range txn.Txn.Accounts {
+		accounts = append(accounts, addr)
+	}
 
-func (l *localLedger) AppGlobalState(appIdx basics.AppIndex) (basics.TealKeyValue, error) {
-	var ap basics.AppParams
+	appIdx := txn.Txn.ApplicationID
 	if appIdx == 0 {
-		if l.groupIndex >= len(l.txnGroup) {
-			return basics.TealKeyValue{}, fmt.Errorf("can't resolve application %d", appIdx)
-		}
-		appIdx = l.txnGroup[l.groupIndex].Txn.ApplicationID
+		// presumably this is app create transaction, initialize with some value
+		appIdx = defaultNewAppIdx
 	}
 
-	allowed := false
-	if appIdx == l.txnGroup[l.groupIndex].Txn.ApplicationID {
-		allowed = true
-	} else {
-		for _, faIdx := range l.txnGroup[l.groupIndex].Txn.ForeignApps {
-			if appIdx == faIdx {
-				allowed = true
-				break
-			}
-		}
+	apps := []basics.AppIndex{appIdx}
+	for _, aid := range txn.Txn.ForeignApps {
+		apps = append(apps, aid)
 	}
-	if !allowed {
-		return basics.TealKeyValue{}, fmt.Errorf("access to the app forbidden %d", appIdx)
+
+	ba := &balancesAdapter{
+		balances:   balances,
+		txnGroup:   txnGroup,
+		groupIndex: groupIndex,
+		proto:      proto,
+		round:      round,
 	}
 
-	for _, br := range l.balances {
-		var ok bool
-		ap, ok = br.AppParams[appIdx]
-		if ok {
-			return ap.GlobalState, nil
-		}
+	return ledger.MakeDebugAppLedger(ba, accounts, apps, appIdx)
+}
 
+func (ba *balancesAdapter) Get(addr basics.Address, withPendingRewards bool) (basics.BalanceRecord, error) {
+	br, ok := ba.balances[addr]
+	if !ok {
+		return basics.BalanceRecord{}, fmt.Errorf("no such address %s", addr.String())
```

Comment:
I believe this function returns an empty balance record and nil error when the address does not exist.

Reply:
done this and below